generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -0,0 +1,102 @@
+package chipyard
+
+import chisel3._
+
+import scala.collection.mutable.{ArrayBuffer}
+
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import freechips.rocketchip.util.{ResetCatchAndSync}
+import chipyard.config.ConfigValName._
+import chipyard.iobinders.{IOBinders, TestHarnessFunction, IOBinderTuple}
+
+import barstools.iocell.chisel._
+
+case object BuildSystem extends Field[Parameters => RawModule]((p: Parameters) => Module(LazyModule(new DigitalTop()(p)).suggestName("system").module))
+
+/**
+ * The base class used for building chips. This constructor instantiates a module specified by the BuildSystem parameter,
+ * named "system", which is an instance of DigitalTop by default. The default clock and reset for "system" are set by two
+ * wires, "systemClock" and "systemReset", which are intended to be driven by traits mixed-in with this base class.
+ */
+abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with HasTestHarnessFunctions {
+
+  // A publicly accessible list of IO cells (useful for a floorplanning tool, for example)
+  val iocells = ArrayBuffer.empty[IOCell]
+  // A list of functions to call in the test harness
+  val harnessFunctions = ArrayBuffer.empty[TestHarnessFunction]
+  // The system clock
+  val systemClock = Wire(Input(Clock()))
+  // The system reset (synchronous to clock)
+  val systemReset = Wire(Input(Bool()))
+
+  // The system module specified by BuildSystem
+  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+
+  // Call all of the IOBinders and provide them with a default clock and reset
+  withClockAndReset(systemClock, systemReset) {
+    val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.map(_(system)).flatten.unzip3
```

Comment:
In the future,  you can replace this map().flatten with flatMap. I've done this in my bump.

Reply:
Ah, yep. Cool